![Cabec%CC%A7alho_notebook.png](cabecalho_notebook.png)

# PCA - Tarefa 01: *HAR* com PCA

Vamos trabalhar com a base da demonstração feita em aula, mas vamos explorar um pouco melhor como é o desempenho da árvore variando o número de componentes principais.

In [28]:
import pandas as pd

from sklearn.tree import DecisionTreeClassifier

from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

In [39]:



filename_features = r"C:\Users\gerar\aula dados\aula1\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\features_info.txt"

filename_labels =r"C:\Users\gerar\aula dados\aula1\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\activity_labels.txt"

filename_subtrain = r"C:\Users\gerar\aula dados\aula1\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\train\subject_train.txt"
filename_xtrain = r"C:\Users\gerar\aula dados\aula1\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\train\X_train.txt"
filename_ytrain = r"C:\Users\gerar\aula dados\aula1\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\train\y_train.txt"

filename_subtest = r"C:\Users\gerar\aula dados\aula1\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\test\subject_test.txt"
ffilename_xtest = r"C:\Users\gerar\aula dados\aula1\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\test\X_test.txt"
filename_ytest = r"C:\Users\gerar\aula dados\aula1\human+activity+recognition+using+smartphones\UCI HAR Dataset\UCI HAR Dataset\test\y_test.txt"

features = pd.read_csv(filename_features, header=None, names=['nome_var'], sep="#")
Names = features['nome_var'].tolist() 
labels = pd.read_csv(filename_labels, delim_whitespace=True, header=None, names=['cod_label', 'label'])

subject_train = pd.read_csv(filename_subtrain, header=None, names=['subject_id'], )
X_train = pd.read_csv(filename_xtrain, delim_whitespace=True, header=None, names=Names)

y_train = pd.read_csv(filename_ytrain, header=None, names=['cod_label'])

subject_test = pd.read_csv(filename_subtest, header=None, names=['subject_id'],)
X_test = pd.read_csv(ffilename_xtest, delim_whitespace=True, header=None, names=Names)
y_test = pd.read_csv(filename_ytest, header=None, names=['cod_label'])

C:\Users\gerar\AppData\Local\Temp\ipykernel_16592\3124692718.py:15: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  labels = pd.read_csv(filename_labels, delim_whitespace=True, header=None, names=['cod_label', 'label'])
C:\Users\gerar\AppData\Local\Temp\ipykernel_16592\3124692718.py:18: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  X_train = pd.read_csv(filename_xtrain, delim_whitespace=True, header=None, names=Names)
C:\Users\gerar\AppData\Local\Temp\ipykernel_16592\3124692718.py:23: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  X_test = pd.read_csv(ffilename_xtest, delim_whitespace=True, header=None, names=Names)


In [40]:
#X_train, X_valid, y_train, y_valid = train_test_split(X_train, y_train)

## Árvore de decisão

Rode uma árvore de decisão com todas as variáveis, utilizando o ```ccp_alpha=0.001```. Avalie a acurácia nas bases de treinamento e teste. Avalie o tempo de processamento.

In [41]:
%%time
clf = DecisionTreeClassifier(random_state=1234, ccp_alpha=0.001).fit(X_train, y_train)

train_scores = clf.score(X_train, y_train)
#valid_scores = clf.score(X_valid, y_valid)
test_scores = clf.score(X_test, y_test)
print(train_scores)
print(test_scores)

0.9390642002176278
0.8031896844248388
CPU times: total: 609 ms
Wall time: 611 ms


## Árvore com PCA

Faça uma análise de componemtes principais das variáveis originais. Utilize apenas uma componente. Faça uma árvore de decisão com esta componente como variável explicativa.

- Avalie a acurácia nas bases de treinamento e teste
- Avalie o tempo de processamento

In [ ]:
list = None

In [75]:
%%time

prcomp = PCA().fit(X_train)

pc_trein= prcomp.transform(X_train)
pc_test= prcomp.transform(X_test)

n=49

colunas = ['cp'+str(x+1) for x in range(n)]

pc_train = pd.DataFrame(pc_trein[:,:n], columns = colunas)
pc_test  = pd.DataFrame( pc_test[:,:n], columns = colunas)



clf = DecisionTreeClassifier(random_state=1234, ccp_alpha=0.001).fit(pc_train, y_train)

train_scores = clf.score(pc_train, y_train)
#valid_scores = clf.score(X_valid, y_valid)
test_scores = clf.score(pc_test, y_test)
print(train_scores)
print(test_scores)



0.8533732317736671
0.7295554801493044
CPU times: total: 1.14 s
Wall time: 1.04 s


## Testando o número de componentes

Com base no código acima, teste a árvore de classificação com pelo menos as seguintes possibilidades de quantidades de componentes: ```[1, 2, 5, 10, 50]```. Avalie para cada uma delas:

- Acurácia nas bases de treino e teste
- Tempo de processamento


# teste n=1 
>> score treino 53%
>> score teste 50%
>> wall time 145ms
# teste n=2
>> score treino 56%
>> score teste 52%
>> wall time 477ms
# teste n=5
>> score treino 77%
>> score teste 71%
>> wall time 319ms
# teste n=10
>> score treino 78%
>> score teste 72%
>> wall time 370ms
# teste n=49
>> score treino 85%
>> score teste 72%
>> wall time 1,04 s

## Conclua

- O que aconteceu com a acurácia?
- O que aconteceu com o tempo de processamento?

>> acuracia aumentou na base de treino, e tambem na base de teste
>> tempo de processamento foi gradativamente subindo.